app/Exceptions/Handler.php

```diff
@@ -2,83 +2,213 @@
 	 *
-	 * @param Request   $request
-	 * @param Throwable $exception
+	 * As we have our own home-brewed logging mechanism via {@link Logs}
+	 * which does not implement {@link \Psr\Log\LoggerInterface} and does
+	 * not register with the service container, we override the method.
+	 */
+	public function report(\Throwable $e): void
+	{
+		$e = $this->mapException($e);
+
+		if ($this->shouldntReport($e)) {
+			return;
+		}
+
+		// We use the severity of the first exception for all subsequent
+		// exceptions, because a causing exception should never be reported
+		// with a higher severity than the eventual exception
+		$severity = self::getLogSeverity($e);
+
+		do {
+			$cause = $this->findCause($e);
+
+			if ($e->getPrevious() !== null) {
+				Logs::log($severity, $cause['method'], $cause['line'], $e->getMessage() . '; caused by');
+			} else {
+				Logs::log($severity, $cause['method'], $cause['line'], $e->getMessage());
+			}
+		} while ($e = $e->getPrevious());
+	}
+
+	public static function getLogSeverity(\Throwable $e): int
+	{
+		return array_key_exists(get_class($e), self::EXCEPTION2SEVERITY) ?
+			self::EXCEPTION2SEVERITY[get_class($e)] :
+			Logs::SEVERITY_ERROR;
+	}
+
+	/**
+	 * Returns the cause of an exception.
 	 *
-	 * @return Response
+	 * It finds the first (most inner) method of Lychee code base which
+	 * caused the exception and returns the method name, the file name and
+	 * the line number.
 	 *
-	 * @throws Throwable
+	 * The backtrace reported by PHP is oddly strange.
+	 * The attribute pair file/line on the one hand-side and class/function
+	 * on the other hand-side of a standard PHP backtrace are off-by-one.
+	 * The reported file/line of an entry of the backtrace don't refer to
+	 * the position *inside* the reported class/function, but where
+	 * class/method has been invoked.
+	 * In particular, if one wants to know the position where the
+	 * exception has been thrown, then one must not look up
+	 * `backtrace[0]['file']` and `backtrace[0]['line']`, resp., but
+	 * use `getFile` and `getLine()` of the exception.
+	 *
+	 * @param \Throwable $e
+	 *
+	 * @return array{file: string, line: int, method: string}
 	 */
-	public function render($request, Throwable $exception): Response
+	private function findCause(\Throwable $e): array
 	{
-		$checks = [];
-		$checks[] = new NoEncryptionKey();
-		$checks[] = new InvalidPayload();
-		$checks[] = new AccessDBDenied();
-		$checks[] = new ApplyComposer();
-
-		foreach ($checks as $check) {
-			if ($check->check($request, $exception)) {
-				// @codeCoverageIgnoreStart
-				return $check->go();
-				// @codeCoverageIgnoreEnd
+		$backtrace = $e->getTrace();
+		$file = $e->getFile();
+		$line = $e->getLine();
+		$class = null;
+		$function = null;
+		foreach ($backtrace as $bt) {
+			$class = $bt['class'] ?? null;
+			$function = $bt['function'] ?? null;
+			if (str_contains($file, $this->appPath)) {
+				break;
 			}
+			$file = $bt['file'];
+			$line = $bt['line'];
```

Comment:
I again encountered a case when these two lines are failing. When testing #1274, I decided to change the permissions on `public/uploads/tracks` directory to make the writing of the track file fail so that I could verify that my simplified front-end error handling code worked as expected.

Instead of an expected message, I got `array key "file" not found`. I worked it around by adding `array_key_exists` tests here. I don't know what the correct fix is, but... we need it!

Reply:
Totally agreed. I will try to trigger the error and then have a look at why the attributes `file` and `line` are missing. They definitely shouldn't, because the PHP docs require every `\Throwable` to have them. 

So all I have to do is to set wrong permissions on one of the upload folders and upload a file, right?

Reply:
Well, if I change the permissions on `big` it doesn't fail here, but it does for `tracks`...

Reply:
> Well, if I change the permissions on `big` it doesn't fail here, but it does for `tracks`...

The joys of inconsistency of PHP. :rofl: The way how the error is reported depends on whether one call `fopen` with parameter `wb` (means write binary) or `w+b` (means write/read binary) even though the cause of the error, insufficient permissions, is the same.

Luckily, I own a proper IDE which allows step-debugging and annotates all variables with their value at runtime. This makes debugging so much easier.

![Screenshot_20220416_123558](https://user-images.githubusercontent.com/7428426/163674445-30ce0bee-3c11-4c04-bfd4-bbf1c5b61e0b.png)

I fixed the problem in: https://github.com/LycheeOrg/Lychee/pull/1217/commits/077bc67e3c0a6c064fbd41034095aa647157ff09

Most of the time went into writing that long comment. Sometimes PHP can be really annoying.